clippy_lints/src/borrow_deref_ref.rs

```diff
@@ -0,0 +1,132 @@
+use crate::reference::DEREF_ADDROF;
+use clippy_utils::diagnostics::{span_lint_and_help, span_lint_and_then};
+use clippy_utils::is_lint_allowed;
+use clippy_utils::source::snippet_opt;
+use clippy_utils::ty::implements_trait;
+use rustc_hir::{ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::Mutability;
+use rustc_middle::ty;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `&*(&T)`.
+    ///
+    /// ### Why is this bad?
+    /// When people deref on an immutable reference `&T`, they may expect return `&U`.
+    /// Accutually `&* (&T)` is still `&T`.
+    /// if you want to deref explicitly, `&** (&T)` is what you need.
+    /// If you want to reborrow, `&T` is enough (`&T` is Copy).
+    ///
+    /// ### Known problems
+    /// false negative on such code:
+    /// ```
+    /// let x = &12;
+    /// let addr_x = &x as *const _ as usize;
+    /// let addr_y = &&*x as *const _ as usize; // assert ok now, and lint triggerd.
+    ///                                         // But if we apply it, it will assert fail.
+    /// assert_ne!(addr_x, addr_y);
+    /// ```
+    ///
+    /// ### Example
+    /// ```rust
+    /// let s = &String::new();
+    ///
+    /// // Bad
+    /// let a: &String = &* s;
+    /// foo(&*s);
+    ///
+    /// // Good
+    /// let a: &String = s;
+    /// foo(&**s);
+    ///
+    /// fn foo(_: &str){ }
+    /// ```
+    pub BORROW_DEREF_REF,
+    complexity,
+    "deref on an immutable reference returns the same type as itself"
+}
+
+declare_lint_pass!(BorrowDerefRef => [BORROW_DEREF_REF]);
+
+impl LateLintPass<'_> for BorrowDerefRef {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, e: &'tcx rustc_hir::Expr<'_>) {
+        if_chain! {
+            if !e.span.from_expansion();
+            if let ExprKind::AddrOf(_, Mutability::Not, addrof_expr) = e.kind;
+            if let ExprKind::Unary(UnOp::Deref, deref_expr) = addrof_expr.kind;
+            if !matches!(deref_expr.kind, ExprKind::Unary(UnOp::Deref, ..) );
+            let ref_ty = cx.typeck_results().expr_ty(deref_expr);
+            if let ty::Ref(_, inner_ty, Mutability::Not) = ref_ty.kind();
+            then{
+
+                let map = cx.tcx.hir();
+                if let Some(parent_hir_id) = map.find_parent_node(e.hir_id){
+                    let span = map.span(parent_hir_id);
+                    if span.from_expansion() {
```

Comment:
You don't need to check the span here. This can only happen when the address-of deref expression is an argument to a macro and that argument is then expanded within an expression created by the macro. Whether or not the expression is a macro argument doesn't change whether you want to lint or not. Only whether a mutable reference is taken will do that.

Removing this check here will require you to add it to the check for `DEREF_ADDROF`

Reply:
You are right. But I  don't know how to deal with this example, without checking parent:
```
// no other lints (deref_addrof) triggered.
fn main() {
    let x=&&12;

    println!("{}",*x); // will lint on
    println!("{}",x);  // remove * is ok

    println!("{:p}",*x); // will lint on
    println!("{:p}",x);  // remove *, the result is different !
}
```

Reply:
Really neither of those cases should be linted as the borrow occurs inside a macro expansion. This should have been fixed by the earlier change to make sure neither the deref nor the borrow expressions were from a macro expansion.

Reply:
I'm sorry I didn't get your meaning. 
I try in 082ae7bca3372d22dc985004331a33b00de950c6 , but it still doesn't work.

Reply:
Looks like `format_args` is broken here. That macro borrows every argument, but the span it gives to that borrow is the same as the argument's span, rather than a span indicating it's from a macro. This should probably be changed on rustc's side.

Reply:
lo and behold I'm already trying to do that https://github.com/rust-lang/rust/pull/90131